### ISSUE TITLE
Unexpected equality issue

### ISSUE THREAD

**Issue by jpellegrini** [egallesio/STklos#655]
Hi @egallesio !

Look:

```scheme
stklos> (= 4999999999999999727876154935214080.0 5000000000000000000000000000000000)
#t
```

All other Scheme implementations return `#f`... I suppose this happens in `do_compare` in `number.c`, but hadn't time to check yet. i'll take a look later.

**Comment by egallesio** [comment]
It seems to be when converting a bignum to a real 

```scheme
stklos> (real-precision 50)
stklos> #i5000000000000000000000000000000000
4999999999999999727876154935214080.0
```

I've tried to spot the problem, but didn't see where it was.

**Comment by jpellegrini** [comment]
Aaah, bit that bigint is nota representable as double. It is changed when converted.

**Comment by jpellegrini** [comment]
Yes, that is indeed the problem:
```scheme
stklos> (real-precision 50)
stklos> (- 5000000000000000000000000000000000 4999999999999999727876154935214080.0)
0.0
```

**Comment by jpellegrini** [comment]
> ```scheme
> stklos> (- 5000000000000000000000000000000000 4999999999999999727876154935214080.0)
> 0.0
> ```

That subtraction gives the same result in other implementations, it's impossible to do otherwise. But... The equality is the problem.

So equality is decided using `sub2(x,y)`...

@egallesio if I understood correctly, the conversion method for `real --- bignum` is to turn the bignum into real, then operate. But we may loose precision there and I'm not sure we can avoid that.

So what about this:

- Do the usual method. If the numbers are different, then return `#f`
- If the usual method returns `#t`, double-check:  if one number is exact and one is inexact, then see if the inexact does represent an integer perfectly. If it does, then they are indeed equal. Otherwise, return `#f`

That would be one extra check.

What do you think?

**Comment by jpellegrini** [comment]
See PR #657

**Comment by egallesio** [comment]
Issue solved with PR #657
